File: discord-bot-fleet/bot-python/crypto.py
"""AES-256-GCM decryption — compatible with Node's crypto.ts encryptString().

Node format: enc:iv_hex:tag_hex:ciphertext_hex
Python's cryptography lib expects ciphertext+tag concatenated for AESGCM.decrypt().
"""

import os
import sys
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt

SALT = b'discord-bot-fleet-v1-salt'  # must match src/crypto.ts
DATA_DIR = os.environ.get('DATA_DIR', os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'data'))
KEY_FILE = os.path.join(DATA_DIR, '.master-key')

_master_key_cache = None


def get_master_key() -> bytes:
    global _master_key_cache
    if _master_key_cache is not None:
        return _master_key_cache

    master_key_env = os.environ.get('MASTER_KEY')
    if master_key_env:
        kdf = Scrypt(salt=SALT, length=32, n=2 ** 14, r=8, p=1)
        _master_key_cache = kdf.derive(master_key_env.encode())
        return _master_key_cache

    # Dev fallback — read from file (must match Node's fallback format)
    if os.path.exists(KEY_FILE):
        with open(KEY_FILE, 'r') as f:
            _master_key_cache = bytes.fromhex(f.read().strip())
        return _master_key_cache

    # Generate new dev key (Node will pick this up via its own fallback)
    os.makedirs(DATA_DIR, exist_ok=True)
    _master_key_cache = AESGCM.generate_key(bit_length=256)
    with open(KEY_FILE, 'w') as f:
        f.write(_master_key_cache.hex())
    try:
        os.chmod(KEY_FILE, 0o600)
    except Exception:
        pass
    print(f'[crypto] WARNING: no MASTER_KEY env, generated dev key at {KEY_FILE}', file=sys.stderr)
    return _master_key_cache


def decrypt_string(payload: str) -> str:
    if not payload or not payload.startswith('enc:'):
        return payload  # not encrypted (legacy plaintext)
    parts = payload.split(':')
    if len(parts) != 4:
        raise ValueError('invalid encrypted payload format')
    iv = bytes.fromhex(parts[1])
    tag = bytes.fromhex(parts[2])
    data = bytes.fromhex(parts[3])
    key = get_master_key()
    aesgcm = AESGCM(key)
    # cryptography lib expects ciphertext + tag concatenated
    plaintext = aesgcm.decrypt(iv, data + tag, None)
    return plaintext.decode('utf-8')


def mask_token(token: str) -> str:
    if not token or len(token) <= 16:
        return '••••' if token else ''
    return f'{token[:10]}...{token[-4:]}'

File: src/lib/crypto.ts
// AES-256-GCM encryption for bot tokens at rest.
// Master key from MASTER_KEY env var (or compiled-in fallback in env.ts).
// In dev with neither set, falls back to a random key persisted to db/.master-key.

import { randomBytes, createCipheriv, createDecipheriv, scryptSync } from 'crypto';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import { join } from 'path';
import { MASTER_KEY } from './env';

const DB_DIR = join(process.cwd(), 'db');
const KEY_FILE = join(DB_DIR, '.master-key');
const SALT = 'discord-bot-fleet-v1-salt';

function getMasterKey(): Buffer {
  // Use the env.ts value (process.env.MASTER_KEY or compiled fallback).
  // IMPORTANT: this MUST be the same key used when tokens were encrypted,
  // otherwise decryptString() will throw on every bot start.
  if (MASTER_KEY) {
    return scryptSync(MASTER_KEY, SALT, 32);
  }
  // Dev fallback — random key persisted to file
  if (!existsSync(KEY_FILE)) {
    if (!existsSync(DB_DIR)) mkdirSync(DB_DIR, { recursive: true });
    const key = randomBytes(32);
    writeFileSync(KEY_FILE, key.toString('hex'), { mode: 0o600 });
    console.warn('[crypto] WARNING: no MASTER_KEY env, generated dev key at', KEY_FILE);
    return key;
  }
  return Buffer.from(readFileSync(KEY_FILE, 'utf8').trim(), 'hex');
}

const ALGO = 'aes-256-gcm';

export function encryptString(plaintext: string): string {
  const key = getMasterKey();
  const iv = randomBytes(12);
  const cipher = createCipheriv(ALGO, key, iv);
  const enc = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return `enc:${iv.toString('hex')}:${tag.toString('hex')}:${enc.toString('hex')}`;
}

export function decryptString(payload: string): string {
  if (!payload || !payload.startsWith('enc:')) return payload; // not encrypted
  const parts = payload.split(':');
  if (parts.length !== 4) throw new Error('invalid encrypted payload format');
  const iv = Buffer.from(parts[1], 'hex');
  const tag = Buffer.from(parts[2], 'hex');
  const data = Buffer.from(parts[3], 'hex');
  const key = getMasterKey();
  const decipher = createDecipheriv(ALGO, key, iv);
  decipher.setAuthTag(tag);
  const dec = Buffer.concat([decipher.update(data), decipher.final()]);
  return dec.toString('utf8');
}

export function maskToken(token: string): string {
  if (!token || token.length <= 16) return '••••';
  return `${token.slice(0, 10)}...${token.slice(-4)}`;
}
